common4j/src/main/com/microsoft/identity/common/java/authorities/NativeAuthCIAMAuthority.kt

```diff
@@ -0,0 +1,116 @@
+//  Copyright (c) Microsoft Corporation.
+//  All rights reserved.
+//
+//  This code is licensed under the MIT License.
+//
+//  Permission is hereby granted, free of charge, to any person obtaining a copy
+//  of this software and associated documentation files(the "Software"), to deal
+//  in the Software without restriction, including without limitation the rights
+//  to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+//  copies of the Software, and to permit persons to whom the Software is
+//  furnished to do so, subject to the following conditions :
+//
+//  The above copyright notice and this permission notice shall be included in
+//  all copies or substantial portions of the Software.
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+
+package com.microsoft.identity.common.java.authorities
+
+import com.microsoft.identity.common.java.exception.ClientException
+import com.microsoft.identity.common.java.logging.LogSession
+import com.microsoft.identity.common.java.logging.Logger
+import com.microsoft.identity.common.java.providers.nativeauth.NativeAuthConstants
+import com.microsoft.identity.common.java.providers.nativeauth.NativeAuthOAuth2Configuration
+import com.microsoft.identity.common.java.providers.nativeauth.NativeAuthOAuth2Strategy
+import com.microsoft.identity.common.java.providers.nativeauth.NativeAuthOAuth2StrategyFactory
+import com.microsoft.identity.common.java.providers.oauth2.OAuth2StrategyParameters
+
+/**
+ * NativeAuthCIAMAuthority class creates a custom authority for native auth flows.
+ */
+class NativeAuthCIAMAuthority (
+    private val authorityUrl: String,
+    val clientId: String
+) : CIAMAuthority(authorityUrl) {
+    companion object {
+        private val TAG = NativeAuthCIAMAuthority::class.java.simpleName
+        //This parameter ensures that authorization endpoint is not fetched from OpenId Configuration.
+        private const val NATIVE_AUTH_USE_OPENID_CONFIGURATION = false
+
+        @Throws(Exception::class)
+        fun getAuthorityFromAuthorityUrl(authorityUrl: String, clientId: String):
+                NativeAuthCIAMAuthority {
+            // Piggy back on the existing authority creation to improve reliability.
```

Comment:
Per the comment it seems that if you wouldn't piggy back on that then reliability would be bad. Is there a reason for that?

Reply:
I'm not sure I understand the question @shahzaibj .
We're piggy-backing on the existing `Authority.getAuthorityFromAuthorityUrl(authorityUrl)`, so that 1) we're not duplicating code, and 2) we're relying on existing (proven) functionality (i.e. making it more reliable).

However, this function currently seems unused except for in tests. @fadidurah  you wrote most of the code in this class, can you help out here?